src/transport/mod.rs

```diff
@@ -247,58 +250,82 @@ pub trait FidoDevice: HIDDevice {
         }
     }
 
-    fn establish_shared_secret(&mut self) -> Result<(SharedSecret, AuthenticatorInfo), HIDError> {
+    fn establish_shared_secret(&mut self) -> Result<SharedSecret, HIDError> {
         if !self.supports_ctap2() {
             return Err(HIDError::UnsupportedCommand);
         }
 
-        let info = if let Some(authenticator_info) = self.get_authenticator_info().cloned() {
-            authenticator_info
-        } else {
-            // We should already have it, since it is queried upon `init()`, but just to be safe
-            let info_command = GetInfo::default();
-            let info = self.send_cbor(&info_command)?;
-            debug!("infos: {:?}", info);
-
-            self.set_authenticator_info(info.clone());
-            info
-        };
+        let info = self
+            .get_authenticator_info()
+            .ok_or(HIDError::DeviceNotInitialized)?;
+        let pin_protocol = PinUvAuthProtocol::try_from(info)?;
 
         // Not reusing the shared secret here, if it exists, since we might start again
         // with a different PIN (e.g. if the last one was wrong)
-        let pin_command = GetKeyAgreement::new(&info)?;
+        let pin_command = GetKeyAgreement::new(pin_protocol);
         let device_key_agreement = self.send_cbor(&pin_command)?;
         let shared_secret = device_key_agreement.shared_secret()?;
         self.set_shared_secret(shared_secret.clone());
-        Ok((shared_secret, info))
+        Ok(shared_secret)
+    }
+
+    /// CTAP 2.0-only version:
+    /// "Getting pinUvAuthToken using getPinToken (superseded)"
+    fn get_pin_token(&mut self, pin: &Option<Pin>) -> Result<PinUvAuthToken, HIDError> {
```

Comment:
I'm not yet sure, I agree on that one.
Keep in mind, that we need to establish a `pin_token` for more commands than just `MakeCredential` and `GetAssertion` (even if they are not yet implemented, like `authenticatorBioEnrollment` or `largeBlobs`).
Those commands calculate `pin_param` differently too, so we would need to cram that all into `determine_pin_uv_auth()`. Possible, but not sure yet if sensible/readable. 
And none of those functions, currently part of `PinUvAuthCommand`
```
    fn client_data_hash(&self) -> ClientDataHash;
    fn set_uv_option(&mut self, uv: Option<bool>);
    fn get_uv_option(&mut self) -> Option<bool>;
    fn get_rp_id(&self) -> Option<&String>;
```
would make sense for those either.

The device can `establish_shared_secret`, `block_and_blink` and `get_pin_token` on its own, without the knowledge of the Request its actually used for. Hence, why I put it there.

But I'm also not married to that location, I just currently don't see the problem.

Reply:
OK, I think the core issue is that `PinUvAuthCommand::determine_pin_uv_auth` mixes a lot of logic for getting a `pinUvAuthToken` (which is command independent) and getting a `pinUvAuthParam` (which is command dependent). The name doesn't help, and there's even a confusing `debug!` statement in `StateMachineCtap2::determine_pin_auth` that makes it seem like we're getting a `pinUvAuthToken` instead of a `Param`.

So maybe `PinUvAuthCommand::determine_pin_uv_auth` should be called `PinUvAuthCommand::determine_pin_uv_auth_param` and should not have a default implementation.

If we do that, then I don't really care if the logic for getting `pinUvAuthToken` goes into `FidoDevice` or into `StateMachineCtap2`.

Reply:
Ok, I added another commit (as a new commit, because you may hate it, so it's easier to revert :D ).
1. Removing `PinUvAuthCommand::determine_pin_uv_auth()` completely, making it a trait with no default implementations now.
2. Adding one more function to `PinUvAuthCommand` that covers the only case where MakeCredential and GetAssertion differ, moving that to the individual implementations.
3. Renaming `PinUvAuthCommand` to make sure its only meant for Mc- and Ga-requests.
4. Moving the functionality into `StateMachineCtap2::determine_pin_uv_auth()`
5. Splitting the error-checks out of the loop (we don't need to check for every PIN-retry, if the device is capable of CTAP2)
6. A lot more renaming, with _hopefully_ clearer names (naming is hard).